### ISSUE TITLE
Haraka, plugin watch not working

### ISSUE THREAD

**Issue by prismanet** [haraka/Haraka#2385]
### system info

Haraka | Haraka.js — Version: 2.8.18
 --- | :---
Node | v8.11.1
OS | Linux smtp-nestor.v7digital.net 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
openssl | OpenSSL 1.0.2k-fips  26 Jan 2017

Running on CentOS 7.x

### Expected behavior
No error message on the Watch web page.

### Observed behavior
I can access the watch web page but there are not data. 

When I let it go for a while, The "Monitor" becomes RED and I get the following on the top right.
[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined[object Event], undefined

watch plugin is enabled in config/plugins
redis plugin is enabled and point to local redis-server
When I check for the existing keys in Redis, I only see karma entries

prompt->redis-cli --scan --pattern '*'
karma|209.171.113.253
karma|185.35.63.177

In the logs, when restarting haraka, there's only two entries connecting to redis
[INFO] [-] [limit] connected to redis://127.0.0.1:6379/4 v3.2.10
[INFO] [-] [redis] connected to redis://127.0.0.1:6379 v3.2.10

As the documentation is pretty thin on set up, I'm opening this ticket if someone can help me out.

Regards,
Sylvain

**Comment by msimerson** [comment]
Hi Sylvian,

Watch gets updates from Haraka via redis pub/sub. If you connect to redis-cli and run the `MONITOR` command, during a SMTP conversation you should see entries like this:

````
1523508756.964675 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"helo.checks\",\"result\":{\"ips\":[\"54.240.13.16\"]}}"
1523508756.964760 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"helo.checks\",\"result\":{\"pass\":\"forward_dns\"}}"
1523508756.964847 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"helo.checks\",\"result\":{\"pass\":\"host_mismatch\"}}"
1523508756.965284 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"karma\",\"result\":{\"score\":2}}"
1523508756.965354 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"karma\",\"result\":{\"awards\":\"131\"}}"
1523508756.967155 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"spf\",\"result\":{\"scope\":\"helo\",\"result\":\"None\",\"domain\":\"a13-16.smtp-out.amazonses.com\",\"emit\":true}}"
1523508756.967737 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"data.uribl\",\"result\":{\"skip\":\"helo unsupported for multi.uribl.com\"}}"
1523508756.967866 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"data.uribl\",\"result\":{\"skip\":\"helo unsupported for multi.surbl.org\"}}"
1523508756.968233 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"data.uribl\",\"result\":{\"pass\":\"helo\"}}"
1523508756.968430 [0 172.16.15.9:57470] "get" "no_tls|54.240.13.16"
1523508756.968600 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"tls\",\"result\":{\"msg\":\"no_tls unset\"}}"
1523508757.205258 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"reset\",\"result\":{\"duration\":1.674}}"
1523508757.205422 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"tls\",\"result\":{\"enabled\":true,\"advertised\":true,\"verified\":false,\"cipher\":{\"name\":\"ECDHE-RSA-AES256-SHA\",\"version\":\"TLSv1/SSLv3\"},\"peerCertificate\":{}}}"
1523508757.237252 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"reset\",\"result\":{\"duration\":1.707}}"
1523508757.237289 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"helo\",\"result\":{\"host\":\"a13-16.smtp-out.amazonses.com\",\"verb\":\"EHLO\"}}"
1523508757.237590 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"access\",\"result\":{\"msg\":\"unlisted(ehlo:any)\"}}"
1523508757.237663 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"helo.checks\",\"result\":{\"multi\":true}}"
1523508757.237934 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"helo.checks\",\"result\":{\"pass\":\"host_mismatch\"}}"
1523508757.238838 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"spf\",\"result\":{\"scope\":\"helo\",\"result\":\"None\",\"domain\":\"a13-16.smtp-out.amazonses.com\",\"emit\":true}}"
1523508757.239346 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"data.uribl\",\"result\":{\"skip\":\"helo unsupported for multi.uribl.com\"}}"
1523508757.239392 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"data.uribl\",\"result\":{\"skip\":\"helo unsupported for multi.surbl.org\"}}"
1523508757.239836 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"data.uribl\",\"result\":{\"pass\":\"helo\"}}"
1523508757.272769 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254" "{\"plugin\":\"reset\",\"result\":{\"duration\":1.742}}"
1523508757.273148 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"access\",\"result\":{\"msg\":\"unlisted(mail)\"}}"
1523508757.273224 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"access\",\"result\":{\"msg\":\"unlisted(mail:any)\"}}"
1523508757.520651 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"mail_from.is_resolvable\",\"result\":{\"pass\":\"has_fwd_dns\"}}"
1523508757.555164 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"spf\",\"result\":{\"scope\":\"mfrom\",\"result\":\"Pass\",\"domain\":\"bounces.amazon.com\",\"emit\":true}}"
1523508757.557648 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"qmail-deliverable\",\"result\":{\"msg\":\"mail_from.not local\"}}"
1523508757.558240 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"data.uribl\",\"result\":{\"skip\":\"envfrom unsupported for multi.uribl.com\"}}"
1523508757.558288 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"data.uribl\",\"result\":{\"skip\":\"envfrom unsupported for multi.surbl.org\"}}"
1523508757.810529 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"data.uribl\",\"result\":{\"pass\":\"envfrom\"}}"
1523508757.811104 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"known-senders\",\"result\":{\"sender\":\"amazon.com\",\"auth\":\"spf\"}}"
1523508757.811314 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"mail_from\",\"result\":{\"action\":\"accept\",\"code\":\"CONT\",\"address\":\"20180412045233efaca55b11be49f8bde288f86d50p0na@bounces.amazon.com\"}}"
1523508757.843693 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"access\",\"result\":{\"msg\":\"unlisted(rcpt)\"}}"
1523508757.843826 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"access\",\"result\":{\"msg\":\"unlisted(rcpt:any)\"}}"
1523508757.846654 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"qmail-deliverable\",\"result\":{\"pass\":\"rcpt.vpopmail dir\",\"emit\":true}}"
1523508757.846938 [0 172.16.15.9:57470] "publish" "result-622B2728-8934-42AE-A6A3-D39C2DCF0254.1" "{\"plugin\":\"known-senders\",\"result\":{\"rcpt_ods\":\"surls.net\"}}"
````

Your web browser downloads `client.js` from Haraka's web server and then establishes a WebSockets connection to Haraka. Over that WebSockets connection, those Redis publish messages are pushed to your web browser, which then formats them and updates the table of connections.

I think the `undefined[object Event]` messages your getting are those events, but something isn't quite right.

**Comment by prismanet** [comment]
Thanks for your help. With the details you gave me I've been able to resolve the issues. The default WebSockets was set to wss (through https which was not set up). For anyone configuring watch, here's what need to be done.

**Requirements**

1. redis plugin enabled in config/plugins
2. watch plugin enabled in config/plugins
3. Create (or copy existing) config/watch.ini
[wss]
url=ws://<server name>/   #for WebSockets through HTTP
url=wss://<server name>/   #for WebSockets through HTTPS

**Troubleshooting**
Enable the developer tools of your browser. As soon as I enabled the Firefox Web Developer Console I saw the errors that Firefox was unable to connect to WebSockets on the specified URL.
